zjit/src/codegen.rs

```diff
@@ -298,6 +301,37 @@ fn gen_insn(cb: &mut CodeBlock, jit: &mut JITState, asm: &mut Assembler, functio
     Some(())
 }
 
+// Get EP at level from CFP
+fn gen_get_ep(asm: &mut Assembler, level: u32) -> Opnd {
+    // Load environment pointer EP from CFP into a register
+    let ep_opnd = Opnd::mem(64, CFP, RUBY_OFFSET_CFP_EP);
+    let mut ep_opnd = asm.load(ep_opnd);
+
+    for _ in 0..level {
+        // Get the previous EP from the current EP
+        // See GET_PREV_EP(ep) macro
+        // VALUE *prev_ep = ((VALUE *)((ep)[VM_ENV_DATA_INDEX_SPECVAL] & ~0x03))
+        let offs = SIZEOF_VALUE_I32 * VM_ENV_DATA_INDEX_SPECVAL;
+        ep_opnd = asm.load(Opnd::mem(64, ep_opnd, offs));
+        ep_opnd = asm.and(ep_opnd, Opnd::Imm(!0x03));
+    }
+
+    ep_opnd
+}
+
+fn gen_nested_getlocal(asm: &mut Assembler, ep_offset: u32, level: NonZeroU32) -> Option<lir::Opnd> {
+    let ep = gen_get_ep(asm, level.get());
+    let offs = -(SIZEOF_VALUE_I32 * i32::try_from(ep_offset).ok()?);
+    Some(asm.load(Opnd::mem(64, ep, offs)))
+}
+
+fn gen_nested_setlocal(asm: &mut Assembler, val: Opnd, ep_offset: u32, level: NonZeroU32) -> Option<()> {
+    let ep = gen_get_ep(asm, level.get());
+    let offs = -(SIZEOF_VALUE_I32 * i32::try_from(ep_offset).ok()?);
```

Comment:
```suggestion
    let offset = -(SIZEOF_VALUE_I32 * i32::try_from(ep_offset).ok()?);
```

Reply:
It's a little confusing that this is also an EP offset but it's different somehow. Maybe there are better names?